### PR TITLE
 Genesis block processing for interoperability module - Closes #7321 

### LINKED ISSUE #7321: Genesis block processing for interoperability module
### Description

Genesis block processing for Genesis state initialization needs to be implemented as described in [LIP0045](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#genesis-state-initialization)

### Acceptance Criteria

- Unit test for all the steps should be there as described in [LIP0045](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#genesis-state-initialization)

### PR DESCRIPTION
**Pull request by Incede**
### What was the problem?

This PR resolves #7321

### How was it solved?

Implement as specified

### How was it tested?

Added unit tests

### CODE REVIEW

File: framework/src/modules/interoperability/utils.ts [hunk 1]
```diff
@@ -593,3 +617,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			throw new Error(
+				`Chain validators store key ${chainValidators.storeKey.toString('hex')} is duplicated.`,
+			);
+		}
+		chainValidatorsStoreKeySet.add(chainValidators.storeKey);
+
+		const { activeValidators, certificateThreshold } = chainValidators.storeValue;
+		if (activeValidators.length < 1 || activeValidators.length > MAX_NUM_VALIDATORS) {
+			throw new Error(
+				' Active validators must have at least 1 element and at most MAX_NUM_VALIDATORS elements.',
```
Comment (ishantiw) [inline framework/src/modules/interoperability/utils.ts:711]:
```suggestion
				`Active validators must have at least 1 element and at most ${MAX_NUM_VALIDATORS} elements.`,
```

File: framework/src/modules/interoperability/utils.ts [hunk 2]
```diff
@@ -593,3 +617,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+	let isAnotherSidechainAccount = 0;
+	for (const chainData of genesisStore.chainDataSubstore) {
+		const chainDataStoreKey = chainData.storeKey;
+		if (chainDataStoreKeySet.has(chainDataStoreKey)) {
+			throw new Error(`Chain data store key ${chainDataStoreKey.toString('hex')} is duplicated.`);
+		}
+		chainDataStoreKeySet.add(chainDataStoreKey);
+
+		const chainAccountStatus = chainData.storeValue.status;
+		if (chainAccountStatus === 2) {
```
Comment (ishantiw) [inline framework/src/modules/interoperability/utils.ts:759]: maybe use constant `CHAIN_TERMINATED`

File: framework/src/modules/interoperability/utils.ts [hunk 3]
```diff
@@ -593,3 +617,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+				tokenID.chainID.equals(ownChainAccount.id)
+			)
+		) {
+			throw new Error(
+				`Chain id corresponding to the channel data store key ${channelData.storeKey.toString(
+					'hex',
+				)} is not valid.`,
+			);
+		}
+		if (tokenID.chainID.readInt32BE(0) === 1 && tokenID.localID.readInt32BE(0) !== 0) {
```
Comment (ishantiw) [inline framework/src/modules/interoperability/utils.ts:683]: use constant `MAINCHAIN_ID`

File: framework/src/modules/interoperability/utils.ts [hunk 4]
```diff
@@ -593,3 +617,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			chainAccountSchema,
+		);
+	}
+
+	if (
+		isAnotherSidechainAccount &&
+		!(chainDataStoreKeySet.has(MAINCHAIN_ID_BUFFER) && chainDataStoreKeySet.has(ownChainAccount.id))
+	) {
+		throw new Error(
+			'If a chain account for another sidechain is present, than a chain account for the mainchain must be present, as well as the own chain account.',
```
Comment (ishantiw) [inline framework/src/modules/interoperability/utils.ts:807]:
```suggestion
			'If a chain account for another sidechain is present, then a chain account for the mainchain must be present, as well as the own chain account.',
```

**Review by ishantiw — COMMENTED** [review]
LGTM, just a few minor comments

File: framework/src/modules/interoperability/utils.ts [hunk 5]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			!outboxRootStoreKeySet.has(chainDataStoreKey) ||
+			!channelDataStoreKeySet.has(chainDataStoreKey) ||
+			!chainValidatorsStoreKeySet.has(chainDataStoreKey)
+		) {
+			throw new Error(
+				`Chain data store key ${chainDataStoreKey.toString(
+					'hex',
+				)} missing in some or all of outbox root, channel data and chain validators stores.`,
+			);
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:785]: For each entry in the chain data substore there must be a corresponding entry (with the same value of storeKey) in the outbox root, channel data, and chain validators substores

File: framework/src/modules/interoperability/utils.ts [hunk 6]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+	}
+
+	const outboxRootStoreKeySet = new dataStructures.BufferSet();
+	const outboxRootStore = getStore(id, STORE_PREFIX_OUTBOX_ROOT);
+	for (const outboxRootData of genesisStore.outboxRootSubstore) {
+		if (outboxRootStoreKeySet.has(outboxRootData.storeKey)) {
+			throw new Error(
+				`Outbox root store key ${outboxRootData.storeKey.toString('hex')} is duplicated.`,
+			);
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:646]: Within each substore, check that all entries have a different storeKey

File: framework/src/modules/interoperability/utils.ts [hunk 7]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+	}
+
+	for (const storeKey of chainValidatorsStoreKeySet) {
+		if (!chainDataStoreKeySet.has(storeKey)) {
+			throw new Error(
+				`Chain validators store key ${storeKey.toString('hex')} is missing in chain data store.`,
+			);
+		}
+	}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:834]: For each entry in the chain data substore there must be a corresponding entry (with the same value of storeKey) in the outbox root, channel data, and chain validators substores and viceversa

File: framework/src/modules/interoperability/utils.ts [hunk 8]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+				!channelDataStoreKeySet.has(chainDataStoreKey) ||
+				!chainValidatorsStoreKeySet.has(chainDataStoreKey)
+			) {
+				throw new Error(
+					`Chain data store key ${chainDataStoreKey.toString(
+						'hex',
+					)} missing in some or all of channel data and chain validators stores.`,
+				);
+			}
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:774]: However, if the chain account has status set to terminated, the entry in the outbox root must not be present.

File: framework/src/modules/interoperability/utils.ts [hunk 9]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+
+	for (const storeKey of terminatedOutboxStoreKeySet) {
+		if (!terminatedStateStoreKeySet.has(storeKey)) {
+			throw new Error(
+				`Terminated outbox store key ${storeKey.toString(
+					'hex',
+				)} missing in terminated state store.`,
+			);
+		}
+	}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:927]: For each entry in the terminated outbox substore there must be a corresponding entry (with the same value of storeKey) in the terminated state substore.

File: framework/src/modules/interoperability/utils.ts [hunk 10]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+
+		const terminatedStateStoreValue = terminatedState.storeValue;
+		if (terminatedStateStoreValue.initialized === false) {
+			if (terminatedOutboxStoreKeySet.has(terminatedStateStoreKey)) {
+				throw new Error(
+					`Uninitialized account associated with terminated state store key ${terminatedStateStoreKey.toString(
+						'hex',
+					)} cannot be present in terminated outbox store.`,
+				);
+			}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:888]: For each entry in the terminated state substore, if the property initialized is set to false, the corresponding entry in the terminated outbox substore must not be present.

File: framework/src/modules/interoperability/utils.ts [hunk 11]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			if (
+				!terminatedStateStoreValue.stateRoot.equals(EMPTY_BYTES) ||
+				terminatedStateStoreValue.mainchainStateRoot?.length !== 32
+			) {
+				throw new Error(
+					`For the uninitialized account associated with terminated state store key ${terminatedStateStoreKey.toString(
+						'hex',
+					)} the stateRoot must be set to empty bytes and mainchainStateRoot to a 32-bytes value.`,
+				);
+			}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:898]: Furthermore, the stateRoot must be set to empty bytes and mainchainStateRoot to a 32-bytes value.

File: framework/src/modules/interoperability/utils.ts [hunk 12]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			if (
+				terminatedStateStoreValue.stateRoot.length !== 32 ||
+				!terminatedStateStoreValue.mainchainStateRoot?.equals(EMPTY_BYTES)
+			) {
+				throw new Error(
+					`For the initialized account associated with terminated state store key ${terminatedStateStoreKey.toString(
+						'hex',
+					)} the mainchainStateRoot must be set to empty bytes and stateRoot to a 32-bytes value.`,
+				);
+			}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:909]: Conversely, if the property initialized is set to true, the stateRoot must be set to a 32-bytes value and mainchainStateRoot to empty bytes.

File: framework/src/modules/interoperability/utils.ts [hunk 13]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			);
+		}
+		chainValidatorsStoreKeySet.add(chainValidators.storeKey);
+
+		const { activeValidators, certificateThreshold } = chainValidators.storeValue;
+		if (activeValidators.length < 1 || activeValidators.length > MAX_NUM_VALIDATORS) {
+			throw new Error(
+				`Active validators must have at least 1 element and at most ${MAX_NUM_VALIDATORS} elements.`,
+			);
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:714]: activeValidators must have at least 1 element and at most MAX_NUM_VALIDATORS elements.

File: framework/src/modules/interoperability/utils.ts [hunk 14]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+		let totalWeight = BigInt(0);
+		for (let j = 0; j < activeValidators.length; j += 1) {
+			const activeValidator = activeValidators[j];
+
+			const { blsKey } = activeValidator;
+			if (j < activeValidators.length - 1 && blsKey.compare(activeValidators[j + 1].blsKey) >= 0) {
+				throw new Error(
+					'Active validators must be ordered lexicographically by blsKey property and pairwise distinct.',
+				);
+			}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:725]: activeValidators must be ordered lexicographically by blsKey property and are pairwise distinct.

File: framework/src/modules/interoperability/utils.ts [hunk 15]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+
+			const { bftWeight } = activeValidator;
+			totalWeight += bftWeight;
+		}
+
+		if (totalWeight > MAX_UINT64) {
+			throw new Error(
+				'The total BFT weight of all active validators has to be less than or equal to MAX_UINT64.',
+			);
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:735]: Let totalWeight be the sum of the bftWeight property of every element in activeValidators. Then totalWeight has to be less than or equal to MAX_UINT64

File: framework/src/modules/interoperability/utils.ts [hunk 16]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+
+		const checkBftWeightValue = totalWeight / BigInt(3) + BigInt(1);
+		if (checkBftWeightValue > totalWeight || checkBftWeightValue > certificateThreshold) {
+			throw new Error('The total BFT weight of all active validators is not valid.');
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:740]: Check that ⌊ ⅓ x totalWeight⌋ + 1 <= certificateThreshold <= totalWeight, where ⌊⋅⌋ is the floor function.

File: framework/src/modules/interoperability/utils.ts [hunk 17]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+	}
+
+	if (
+		isAnotherSidechainAccount &&
+		!(chainDataStoreKeySet.has(MAINCHAIN_ID_BUFFER) && chainDataStoreKeySet.has(ownChainAccount.id))
+	) {
+		throw new Error(
+			'If a chain account for another sidechain is present, then a chain account for the mainchain must be present, as well as the own chain account.',
+		);
+	}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:810]: If a chain account for another sidechain is present, than a chain account for the mainchain (chainID = 1) must be present, as well as the own chain account

File: framework/src/modules/interoperability/utils.ts [hunk 18]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+				)} is not valid.`,
+			);
+		}
+		if (tokenID.chainID.equals(MAINCHAIN_ID_BUFFER) && tokenID.localID.readInt32BE(0) !== 0) {
+			throw new Error(
+				`Local id corresponding to the channel data store key ${channelData.storeKey.toString(
+					'hex',
+				)} is not valid.`,
+			);
+		}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:690]: Either channel.messageFeeTokenID={chainID=1, localID=0} (corresponding to the LSK token), or channel.messageFeeTokenID.chainID must be equal to channelData.storeKey or ownChainAccount.ID (the message fee token must be a native token of either chains).

File: framework/src/modules/interoperability/utils.ts [hunk 19]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+
+		await channelDataStore.setWithSchema(
+			channelData.storeKey,
+			channelData.storeValue,
+			channelSchema,
+		);
+	}
```
Comment (Incede) [inline framework/src/modules/interoperability/utils.ts:697]: For every substore, create all the corresponding entries in the Interoperability module state.

File: framework/src/modules/interoperability/utils.ts [hunk 20]
```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
...
+			if (terminatedOutboxStoreKeySet.has(terminatedStateStoreKey)) {
+				throw new Error(
+					`Uninitialized account associated with terminated state store key ${terminatedStateStoreKey.toString(
+						'hex',
+					)} cannot be present in terminated outbox store.`,
+				);
+			}
+			if (
+				!terminatedStateStoreValue.stateRoot.equals(EMPTY_BYTES) ||
+				terminatedStateStoreValue.mainchainStateRoot?.length !== 32
```
Comment (ishantiw) [inline framework/src/modules/interoperability/utils.ts:891]: I think we shoulduse byteLength
```suggestion
				terminatedStateStoreValue.mainchainStateRoot?.byteLength !== 32
```

File: framework/test/unit/modules/interoperability/utils.spec.ts [hunk 21]
```diff
@@ -1095,4 +1118,1117 @@ describe('Utils', () => {
...
+		let chainDataSubstore: any;
+		let terminatedStateSubstore: any;
+		let chainValidatorsSubstore: any;
+		let ownChainDataSubstore: any;
+		let registeredNamesSubstore: any;
+		let registeredNetworkIDsSubstore: any;
+
+		let mockGetStore: any;
+
+		// eslint-disable-next-line @typescript-eslint/require-await
```
Comment (ishantiw) [inline framework/test/unit/modules/interoperability/utils.spec.ts:1259]: I think we don't need this

**Review by ishantiw — APPROVED** [review]
Just a couple of minor comments, approving from my end 👍

File: framework/src/modules/interoperability/utils.ts (hunk 20 above)
Comment (Incede) [reply framework/src/modules/interoperability/utils.ts:891]: I am not sure why you think this, can you maybe elaborate what's wrong with the current logic ?

Comment (ishantiw) [reply framework/src/modules/interoperability/utils.ts:891]: hmm not sure what was the exact reasoning, seems like both of them are returning the number of bytes in the end. @shuse2 do you remember what was the reason we wanted to use byteLength?

Comment (ishantiw) [reply framework/src/modules/interoperability/utils.ts:891]: we can tackle this point in a separate issue

**Review by mitsuaki-u — APPROVED** [review]

**Pull request merged**